File: leetcode/20-valid-parenthesis.rb
class Stack
  def initialize
    @stack = []
  end

  def push(item)
    @stack.push(item)
  end

  def pop
    @stack.pop
  end

  def peek
    @stack.last
  end

  def empty?
    @stack.empty?
  end

  def size
    @stack.size
  end
end

OPEN_VALUES = ["[","{", "("]
PAIRS = ["[]", "{}", "()"]

# @param {String} s
# @return {Boolean}
def is_valid(s)
  stack = Stack.new

  s.chars.each do |char|
    if OPEN_VALUES.include?(char)
      stack.push(char)
    else
      return false if stack.empty?
      if PAIRS.include?(stack.pop << char)
        next
      else
        return false
      end
    end
  end

  stack.empty?
end

File: leetcode/20-valid-parenthesis.py
class Stack:
  def __init__(self):
    self.stack = []

  def push(self, item):
    self.stack.append(item)

  def pop(self):
    if not self.is_empty():
        return self.stack.pop()
    return None  # Handle empty stack case

  def peek(self):
    return self.stack[-1] if not self.is_empty() else None

  def is_empty(self):
    return len(self.stack) == 0

  def size(self):
    return len(self.stack)

class Solution:
  def isValid(self, s: str) -> bool:
    # Initialize an empty list to use as a stack
    stack = Stack()
    # Create a set with valid parentheses pairs
    valid_pairs = {'()', '[]', '{}'}

    # Iterate over each character in the string
    for char in s:
      # If the character is an opening parenthesis, push it onto the stack
      if char in '({[':
        stack.push(char)
      # If the stack is empty or the formed pair is not valid, return False
      elif not stack or stack.pop() + char not in valid_pairs:
        return False

    # If the stack is empty, all parentheses were valid and correctly nested
    return stack.is_empty()